blockchain/v2/scheduler.go

```diff
@@ -0,0 +1,715 @@
+	}
+
+	if peer.state == peerStateRemoved {
+		return fmt.Errorf("cannot set peer height for a peer in peerStateRemoved")
+	}
+
+	if height < peer.height {
+		return fmt.Errorf("cannot move peer height lower. from %d to %d", peer.height, height)
+	}
+
+	peer.height = height
+	peer.state = peerStateReady
+
+	sc.addNewBlocks()
+	return nil
+}
+
+func (sc *scheduler) getStateAtHeight(height int64) blockState {
+	if height <= sc.initHeight {
+		return blockStateProcessed
+	} else if state, ok := sc.blockStates[height]; ok {
+		return state
+	} else {
+		return blockStateUnknown
+	}
+}
+
+func (sc *scheduler) getPeersAtHeight(height int64) []p2p.ID {
+	peers := make([]p2p.ID, 0)
+	for _, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if peer.height >= height {
+			peers = append(peers, peer.peerID)
+		}
+	}
+	return peers
+}
+
+func (sc *scheduler) peersInactiveSince(duration time.Duration, now time.Time) []p2p.ID {
+	peers := make([]p2p.ID, 0)
+	for _, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if now.Sub(peer.lastTouched) > duration {
+			peers = append(peers, peer.peerID)
+		}
+	}
+	return peers
+}
+
+func (sc *scheduler) peersSlowerThan(minSpeed int64) map[p2p.ID]struct{} {
+	peers := make(map[p2p.ID]struct{})
+	for peerID, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if peer.lastRate < minSpeed {
+			peers[peerID] = struct{}{}
+		}
+	}
+	return peers
+}
+
+// XXX: this duplicates the logic of peersInactiveSince and peersSlowerThan
+func (sc *scheduler) prunablePeers(peerTimout time.Duration, minRecvRate int64, now time.Time) []p2p.ID {
+	prunable := make([]p2p.ID, 0)
+	for peerID, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if now.Sub(peer.lastTouched) > peerTimout || peer.lastRate < minRecvRate {
+			prunable = append(prunable, peerID)
+		}
+	}
+	// Tests for handleTryPrunePeer() may fail without sort due to range non-determinism
+	sort.Sort(PeerByID(prunable))
+	return prunable
+}
+
+func (sc *scheduler) setStateAtHeight(height int64, state blockState) {
+	sc.blockStates[height] = state
+}
+
+func (sc *scheduler) markReceived(peerID p2p.ID, height int64, size int64, now time.Time) error {
+	peer, ok := sc.peers[peerID]
+	if !ok {
+		return fmt.Errorf("couldn't find peer %s", peerID)
+	}
+
+	if peer.state == peerStateRemoved {
+		return fmt.Errorf("cannot receive blocks from removed peer %s", peerID)
+	}
+
+	if state := sc.getStateAtHeight(height); state != blockStatePending || sc.pendingBlocks[height] != peerID {
+		return fmt.Errorf("received block %d from peer %s without being requested", height, peerID)
+	}
+
+	pendingTime, ok := sc.pendingTime[height]
+	if !ok || now.Sub(pendingTime) <= 0 {
+		return fmt.Errorf("clock error: block %d received at %s but requested at %s",
+			height, pendingTime, now)
+	}
+
+	// This may divide by zero - needs to measure every 10s blocks (see v0 and v1),
```

Comment:
`now.Sub(pendingTime) <= 0` above ensures that it's always positive. Outdated comment?

Reply:
Yes I think this is outdated as we specifically check for this case on line `412`